### ISSUE TITLE
Possible bug when using `TypeIs`

### ISSUE THREAD

**Issue by rayansostenes** [microsoft/pyright#7767]
I'm confused about the behavior of `test_function1` and `test_function2` on the code bellow, reading PEP 742, I was expecting that the narrowed type of `a` would be the intersection of `MyBaseClass` and `MyProtocol`, like it does on `test_function3` and `test_function4` but it's `Never`.

am I missing something? or is this a bug in pyright?

```python
# pyright: strict
from typing import Any, Protocol, reveal_type

from typing_extensions import TypeIs


class MyBaseClass: ...


class MyProtocol(Protocol):
    x: bool


def is_my_protocol(value: Any) -> TypeIs[MyProtocol]: ...


def test_function1(a: MyBaseClass):
    if is_my_protocol(a):
        reveal_type(a)  # Type of "a" is "Never"


def test_function2(a: object):
    if isinstance(a, MyBaseClass) and is_my_protocol(a):
        reveal_type(a)  # Type of "a" is "Never"


def test_function3(a: object):
    if is_my_protocol(a) and isinstance(a, MyBaseClass):
        reveal_type(a)  # Type of "a" is "<subclass of MyProtocol and MyBaseClass>"


def test_function4(a: MyProtocol):
    if isinstance(a, MyBaseClass):
        reveal_type(a)  # Type of "a" is "<subclass of MyProtocol and MyBaseClass>"
```
[Pyright Playground](https://pyright-play.net/?pythonVersion=3.12&strict=true&locale=en&code=MQAgDgngTglg5gCwC4C4QGcmwMZIFABmUA9gLYhIRgwB2cIMpYxUSIAgjRADQgAKJJMWzEANrygBTAG6SAhqID6lMJLyES5FbTiLJADySSa6GMRMMmLNgBUqkgJLp1ebKLnp0IALIQAQh6SAMLunmgAdJEubh5evgLEQiKiABQJSWIAlCh4IHkg%2BmgARsRiLgAmkgQM6IqkEIpggsJiKdIKAK6SaJwQmSAAtAB8IHaqTgDa8c3JALoRUeqV1UaYigQdNLhmNACMKXJovgHowaHo2bn5MNUwtfWNM61yl-lvIFKyCsr2B-0goDGkhAxGqACI5GCaiAwQA5GSSKBgipVCiSNYbLZIHYAJgOaGIRQAVpJcK9rrdTCYkHItpIDrxjoEQrF%2BrTyjU6g0mokWqkXjl3vlPvIlCp6S88oD7CDwZDoXCEUiUSt0Uh1pttuYAMz4kHE0lIcl5G6ch48jL8tk0Dl3WiYWnYCWM-zM87G94i77iv5S0Yy0Ew%2BV3GEAHnQHSKMU8sp8EHSfJA7LjJzOsSGyKWqNW6sxWpoABY9dNeckPaa7dTHc6U27WYKhR8Ed7fpKAf7VLGIVCQ2Dw5Ho15AyXLUmbbXTizPBmgA)

**Comment by erictraut** [comment]
The `TypeIs` functionality in pyright doesn't product an intersection currently. There is officially no intersection support in the Python type system today. Both mypy and pyright support a very limited form of intersection when narrowing with `isinstance`, but this logic (at least in pyright's case) is distinct from the narrowing logic used by `TypeIs`.

The `TypeIs` specification gives type checkers latitude in how they implement the functionality since the theoretically-correct types cannot be expressed in today's type system. The result is therefore going to be an approximation. There may be cases where a narrowed type is not as narrow as it could be, but it shouldn't ever be narrower. In your example, narrowing to `Never` is too narrow, so I'd consider this a bug.

**Comment by erictraut** [comment]
I've decided to combine the logic for `TypeIs` and `isinstance` narrowing so they are consistent. That's likely the behavior that most pyright users expect. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.361, which I just published.
